spec/requests/memberships_controller_spec.rb

```diff
@@ -1,130 +1,282 @@
 # frozen_string_literal: true
 
 describe MembershipsController do
+  shared_context 'current_user is the group leader' do
+    let(:group)   { membership.group }
+    before(:each) { login_as(group.leader.user) }
+  end
+
+  shared_context 'current_user is not a member' do
+    let(:group)   { create(:group) }
+    let(:user)    { create(:user) }
+    before(:each) { login_as(user) }
+  end
+
+  shared_context 'current_user is a member' do
+    let(:group)   { membership.group }
+    let(:user)    { membership.user }
+    before(:each) { login_as(user) }
+  end
+
   describe '#create' do
-    it 'when user is not a member creates membership' do
-      user  = create(:user)
-      group = create(:group)
-      login_as(user)
+    context 'when membership is not created and current_user is not a member' do
+      include_context 'current_user is not a member'
 
-      post "/groups/#{group.id}/memberships"
-      user.memberships.reload
+      it 'creates membership' do
+        post "/groups/#{group.id}/memberships"
+        user.memberships.reload
 
-      expect(group.leader.user.notifications.count).to be 1
-      expect(user.membership_for(group)).not_to be_nil
-      expect(subject).to redirect_to group_path(group)
+        expect(group.leader.user.notifications.count).to be 1
+        expect(user.membership_for(group)).not_to be_nil
+        expect(response).to redirect_to group_path(group)
+      end
     end
 
-    it 'when user already member it returns forbidden' do
-      group = create(:group)
-      user  = group.members.first
-      login_as(user)
+    context 'when membership already created and current_user is member' do
+      include_context 'current_user is a member'
+      let(:membership) { create(:membership) }
 
-      post "/groups/#{group.id}/memberships"
+      it 'it returns forbidden' do
+        post "/groups/#{group.id}/memberships"
 
-      expect(Membership.where(group: group, user: user).count).to eq(1)
-      expect(response).to have_http_status(:forbidden)
+        expect(Membership.where(group: group, user: user).count).to eq(1)
+        expect(response).to have_http_status(:forbidden)
+      end
+    end
+
+    context 'when membership is archived and current_user is member' do
+      include_context 'current_user is a member'
+      let(:membership) { create(:membership, :archived) }
+
+      it 'notifies the group leader' do
+        expect_any_instance_of(Membership).to receive(:notify).and_call_original
+
+        expect {
```

Comment:
The group leader already had one notification from creating the membership. Expecting eq(2) on the `notifications.count` seemed  weird so I used this.